File: Collector/src/main/java/com/bdb/weather/collector/CurrentWeatherPublisher.java
/* 
 * Copyright (C) 2015 Bruce Beisel
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.bdb.weather.collector;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.InetAddress;
import java.net.MulticastSocket;
import java.net.UnknownHostException;
import java.util.logging.Level;
import java.util.logging.Logger;

import com.bdb.weather.common.CurrentWeatherSubscriber;

/**
 * Publishes the current weather on a multicast socket. This means that the database is not needed to
 * communicate the current weather to any running user interfaces.
 * 
 * @author Bruce
 */
final class CurrentWeatherPublisher {
    private MulticastSocket socket;
    private int port;
    private InetAddress group;
    private static final Logger logger = Logger.getLogger(CurrentWeatherPublisher.class.getName());
    
    /**
     * Constructor that opens the default address and default port.
     * 
     * @throws UnknownHostException Could not initialize the Multicast socket
     * @throws IOException Could not open the port
     */
    public CurrentWeatherPublisher() throws UnknownHostException, IOException {
        this(CurrentWeatherSubscriber.DEFAULT_ADDRESS, CurrentWeatherSubscriber.DEFAULT_PORT);
    }
    
    /**
     * Constructor.
     * 
     * @param address The address on which to create the Multicast socket
     * @param port The port on which to create the Multicast socket
     * @throws UnknownHostException Could not initialize the Multicast socket
     * @throws IOException Could not open the port
     */
    public CurrentWeatherPublisher(String address, int port) throws IOException, UnknownHostException {
        socket = new MulticastSocket(port);
        socket.setTimeToLive(2);
        group = InetAddress.getByName(address);
        this.port = port;
    }
    
    /**
     * Send the current weather as XML on the Multicast socket.
     * 
     * @param xml The current weather XML
     */
    public void sendCurrentWeather(String xml) {
        try {
            byte[] b = xml.getBytes();
            DatagramPacket packet = new DatagramPacket(b, b.length, group, port);
            socket.send(packet);
        }
        catch (IOException e) {
            logger.log(Level.WARNING, "Exception while publishing current weather", e);
        }
    }
}

File: VantagePro2Driver/VP2++/src/VP2/CurrentWeatherPublisher.cpp
/* 
 * Copyright (C) 2022 Bruce Beisel
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#include <sys/socket.h>
#include <sys/types.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <unistd.h>
#include <ifaddrs.h>
#include <string.h>
#include <string>
#include "VP2Logger.h"
#include "CurrentWeatherPublisher.h"
#include "CurrentWeather.h"

using namespace std;

namespace vp2 {
const std::string CurrentWeatherPublisher::MULTICAST_HOST = "224.0.0.120";

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
CurrentWeatherPublisher::CurrentWeatherPublisher() : log(VP2Logger::getLogger("CurrentWeatherPublisher"))
{
    createSocket();
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
CurrentWeatherPublisher::~CurrentWeatherPublisher()
{
    if (socketId != NO_SOCKET)
        close(socketId);
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
void
CurrentWeatherPublisher::sendCurrentWeather(const CurrentWeather & cw)
{
    if (socketId == NO_SOCKET)
        return;

    std::string s = cw.formatXML();
    const char * data = s.c_str();
    size_t length = strlen(data);
    if (sendto(socketId, data, length, 0, reinterpret_cast<struct sockaddr *>(&groupAddr), sizeof(groupAddr)) != length) {
        int e = errno;
        log.log(VP2Logger::VP2_WARNING) <<  "sendto() for current weather failed. Errno = " << e << endl;
    }
    else
        log.log(VP2Logger::VP2_INFO) << "Published current weather: " << data << endl;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
bool
CurrentWeatherPublisher::getLocalIpAddress(struct sockaddr_in & saddr)
{
    bool rv = false;
    struct ifaddrs *addrs;
    struct ifaddrs *tmp;

    getifaddrs(&addrs);
    tmp = addrs;

    while (tmp) {
        if (tmp->ifa_addr && tmp->ifa_addr->sa_family == AF_INET) {
            struct sockaddr_in *pAddr = reinterpret_cast<struct sockaddr_in *>(tmp->ifa_addr);
            if (strncmp(tmp->ifa_name, "lo", 2) != 0) {
                printf("%s: %s\n", tmp->ifa_name, inet_ntoa(pAddr->sin_addr));
                saddr = *pAddr;
                rv = true;
                break;
            }
        }

        tmp = tmp->ifa_next;
    }

    freeifaddrs(addrs);

    return rv;
}

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
bool
CurrentWeatherPublisher::createSocket()
{
    if ((socketId = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
        socketId = NO_SOCKET;
        return false;
    }

    memset(reinterpret_cast<char *>(&groupAddr), 0, sizeof(groupAddr));
    groupAddr.sin_family = AF_INET;
    groupAddr.sin_addr.s_addr = inet_addr(MULTICAST_HOST.c_str());
    groupAddr.sin_port = htons(MULTICAST_PORT);

    struct sockaddr_in saddr;
    if (!getLocalIpAddress(saddr)) {
        log.log(VP2Logger::VP2_ERROR) <<  "setsockopt() getting local IP address failed." << endl;
        close(socketId);
        socketId = NO_SOCKET;
        return false;
    }

    if (setsockopt(socketId, IPPROTO_IP, IP_MULTICAST_IF, reinterpret_cast<char *>(&saddr.sin_addr), sizeof(saddr.sin_addr)) < 0) {
        int e = errno;
        log.log(VP2Logger::VP2_ERROR) <<  "setsockopt() for local interface failed. Errno = " << e << endl;
        close(socketId);
        socketId = NO_SOCKET;
        return false;
    }

    unsigned char ttl = 2;
    if (setsockopt(socketId, IPPROTO_IP, IP_MULTICAST_TTL, reinterpret_cast<char *>(&ttl), sizeof(ttl)) < 0) {
        int e = errno;
        log.log(VP2Logger::VP2_ERROR) <<  "setsockopt() for TTL failed. Errno = " << e << endl;
        close(socketId);
        socketId = NO_SOCKET;
        return false;
    }

    log.log(VP2Logger::VP2_INFO) << "Multicast socket created successfully" << endl;
    return true;
}

}
